File: src/09_externalfiles.r
library(data.table)
library(jsonlite)
library(readxl)
library(xml2)

setwd("~/sas-py-r/src")

# ---------
# CSV Files
# ---------
policies <- data.table(read.csv("../data/policies.csv"))

# ----------
# JSON Files
# ----------
jsondata <- fromJSON("../data/sample.json")
print(paste("JSON:", jsondata))

# ----------
# XSLX Files
# ----------
characters <- data.table(read_excel("../data/sample.xlsx"))

# ---------
# XML Files
# ---------
xmldata <- read_xml("../data/sample.xml")
xmlcharacters <- xml_text(xml_find_all(xmldata, "//character//name"))
xmlanswer <- as.integer(xml_text(xml_find_all(xmldata, "//answer")))
print(paste("XML: (characters)", xmlcharacters))

# ------------------
# Unstructured Files
# ------------------
lines <- readLines("../data/sample.txt")
for (line in lines) {
  if (length(line) == 0) {
    continue
  } else if (substr(line,0,1) != '-') {
    character <- line
  } else {
    line <- strsplit(line, ",")[[1]]
    actor <- line[1]
    actor <- substr(actor, 3, nchar(actor))
    methods <- line[2:length(line)]
    print(paste(character, "played by", actor, "in", paste(methods, collapse=" &")))
  }
}

File: src/09_externalfiles.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import pandas as pd
import json
from xml.etree import ElementTree
    
# ---------
# CSV Files
# ---------
policies = pd.read_csv("../data/policies.csv")

# ----------
# JSON Files
# ----------
with open("../data/sample.json", 'r') as f:
    jsondata = json.load(f)
print("JSON:", jsondata)

# ----------
# XLSX Files
# ----------
characters = pd.read_excel("../data/sample.xlsx")

# ---------
# XML Files
# ---------
tree = ElementTree.parse("../data/sample.xml")
for child in tree.getroot():
    for subchild in child.iter():
        if subchild.tag in ("name", "answer"):
            print("XML:", subchild.tag, "=", subchild.text)

# ------------------
# Unstructured Files
# ------------------
with open("../data/sample.txt", 'r') as f:
    for line in f.readlines():
        if len(line)==0:
            continue
        elif line[0]!='-':
            character = line.rstrip()
        else:
            line = line[2:].rstrip().split(',')
            actor = line[0]
            methods = [m.strip() for m in line[1:]]
            print(character, "played by", actor, "in", " & ".join(methods))

